resolves #510 add support for projects hosted on GitLab

mojavelinux opened the pull request:
This PR adds support for projects hosted on GitLab, including star count.

~Due to how the GitLab API works, it's necessary to specify the repo using the ID instead of the path. The repo property in the front matter must be defined as follows:~

```
repo: gitlab:4180516
```

UPDATE: This PR now implements a separate repohost field. And we found out that the GitLab API accepts a repo path as well. So now the configuration is:

```
repo: antora/antora
repohost: gitlab
```

The repo will get replaced with the URL returned from the API.

In order to use these changes, it's necessary to:

1. run `npm i` to get the new package
2. set the STATICGEN_GITLAB_TOKEN environment variable to a personal access token
3. remove and rebuild the staticgen-archive.json file

I also found I had to update react to at least 16.3.0 to get the site to build without error. (UPDATE: This can be avoided by installing packages using yarn).

mojavelinux commented on static.config.js:
```diff
@@ -75,6 +75,8 @@ function extractRelevantProjectData (data) {
       followersPrevious,
       dataAgeInDays,
     }
+    if (repo) relevantData.repo = repo
```
Here we are allowing the API to give us the web/html URL of the repository instead of having to hardcode it.

mojavelinux commented:
@erquhart I'm keen to get your feedback on this PR. I'm happy to update if there's something about it that you think could be improved.

mojavelinux commented:
Thanks for the feedback! I'll work on the updates.

> the GitLab API accepts this as the project id if it's URL encoded

Wow, I looked for that for 20 minutes and somehow missed. That will make things a lot easier. I'll give it a shot.

> We shouldn't need to store any additional or different data in the Gist cache

I kind of like the idea of allowing the API to resolve the repo URL instead of hard coding it. If we take that route, doesn't it have to be stored in the cache?

> We should update the "GitHub stars" label(s) in the UI, to something generic, as it isn't accurate as of this PR, maybe "Repo stars".

Oh, good point! I looked right past that one.

> Can you add Antora to this PR so we can see it in action?

I would say yes, except the [other Antora PR](https://github.com/netlify/staticgen/pull/429) is still open and the submitter requested that we built on it. Can we take that route instead?

erquhart commented:
I'm not against allowing the API to provide the URL, but we're not doing it for GitHub, and our templates are still set up for hardcoded url building. I'm thinking if we want to change how the URL is retrieved/built, let's handle it under a separate PR and keep this one straightforward.

Regarding the doc PR, I'm all for honoring community effort, but the PR in question contains a copy paste from the main page of the docs, so there's no real effort being lost. If you'd like to pull her commit into your PR here and then add another to get the page where you want it, please do! We just need this PR to include a working application.

mojavelinux commented:
> We just need this PR to include a working application.

In that case, I'll move forward with including the entry for Antora in the PR. Stay tuned.

mojavelinux commented:
> but we're not doing it for GitHub, and our templates are still set up for hardcoded url building.

My PR had included this change. And the reason for it is that I was trying to avoid adding repository-specific logic to the templates. I think it's better if the UI passes a pure model than to be hard-coding github.com in the templates (and thus the template knows about the repository types). I'm fine with doing it in a separate PR, but it really feels part of this PR to me.

erquhart commented:
Ah, somehow I missed that you addressed the templating issue. I'll take another look when I get to my machine. The only other concern is how we're storing the data. I _think_ each project has any array of cache objects, each with stars, Twitter flowers, etc, for each date. We need to make sure we're not storing the repo url for each date, but rather once per project, and just rewrite it each time in case it changes.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

erquhart commented:
Actually nvm on storing the repo once per project. We _should_ do it that way, but I've already limited the cache to 7 days of data per project, so the redundant data won't put us over the gist size limit. Someone can improve it in the future.

mojavelinux commented:
Thanks for the additional insight. I'm going to try to complete the update tomorrow night. We'll then iterate from there.

mojavelinux commented:
> Given that point 1 makes it impossible to infer the repo host from the repo name, we should explicitly state the repo host in a project's frontmatter

Why not a prefix like `gitlab:antora/antora`?

mojavelinux commented:
The reason I suggest the prefix is because a) it's simpler to type, b) allows the repo field to be globally unique, and c) requires the smallest change to the code (which only passes the repo string in several places).

mojavelinux commented:
@erquhart I have updated the PR.

Btw, I still can't build the site locally without upgrading React. I'm not sure how it's working on Netlify.

mojavelinux commented:
It would be nice if this repo published deploy previews to Netlify, or is that already available?

erquhart commented:
There are plenty of reasons to access the repo host name vs the repo location, they're separate data points. I get the uniqueness thing as generally ideal, but it only has real value if we depend on that uniqueness somewhere. The format `<host>:<user>/<project>` is a bit of a DSL, and the benefits aren't really there. I'd rather see us call these distinct data points what they are explicitly.

Again, I'm with taking on redundant data for expedience, like the repo url being repeated a bunch of times in our data, but I don't think there's much more work to add an optional `repo_host` field that would only need to be added to the Antora project initially.

erquhart commented:
Also, deploy preview is up now, sorry about that 👍

mojavelinux commented:
> Also, deploy preview is up now, sorry about that

It looks like the GitLab integration isn't working in the deploy preview. Perhaps the token isn't set?

mojavelinux commented:
> There are plenty of reasons to access the repo host name vs the repo location, they're separate data points. 

Yes, I get that, but,

1. The site doesn't, currently
2. This introduces a much larger which just isn't needed right now

> I get the uniqueness thing as generally ideal, but it only has real value if we depend on that uniqueness somewhere.

But it *does* depend on that uniqueness because the repo is being used as a key. Changing that means changing how the data is passed around, which involves several functions.

I was trying to keep the change as localized as possible. You seem to want to overhaul the core project logic. I'm fine with doing that, but you need to give me a clear indicator that would be accepted before I spend time on it.

> I don't think there's much more work to add an optional repo_host field that would only need to be added to the Antora project initially

Yes, it is a lot more work. And since I'm the only one that seems to be willing to do it, I'd ask that you trust my advice. It seems to me like the milestone keeps getting moved to have Antora added to this site, which frankly is frustrating me. But I'm willing to stick with it if you can confirm that you are ready for whatever change I have to make to support this repo_host field correctly.

mojavelinux commented:
And, btw, `repo_host` would introduce an inconsistent naming pattern. There's already a frontmatter key named `startertemplaterepo` that follows a different pattern. So do you want the name of the field to be `repohost` instead?

mojavelinux commented:
For the record, I don't see what the problem with `gitlab:antora/antora` as the repo name is. That's a globally unique name and easy to parse. And without the prefix, it's meaningless anyway (because it doesn't tell you what host it is). So I'd argue that the host is part of the key.

erquhart commented:
I wrote up a very well thought up response, and then navigated away in my over-confidence regarding GitHub's draft backup functionality 🙄 

I'll just summarize:
Your work on this is greatly appreciated, I don't take it for granted a bit. There was no intention of moving the goal post, more like the goal post was never set up because I didn't expect to be thinking about this at the moment. Definitely want to help you get Antora added, just want to get it right.

Regarding `repo_host` (or more accurately, `repohost`, thanks for pointing that out), I'm loosely expecting the following:
1. In `scripts/fetch-archive.js`, we gather the `repo_host` and `repo` from the project frontmatters
2. We convert the `repohost`/`repo` pairs into urls from the respective Git host API's
3. We only use the resulting URL across the app, `repo` and `repohost` never leave the script
4. Anywhere outside of `fetch-archive.js` that uses a project's `repo` value as a unique key can _still_ use it as a unique key, as it should be a unique URL returned by a Git provider API

I could easily be missing something, as I haven't touched this codebase much in some time, so please correct any inaccuracies there.

Finally, I'm going to let the above be my last feedback on the approach. You were writing books about programming before I was even in the field, so obviously whatever you submit will be a great solution. If my suggested approach here doesn't fly, just go with your gut, no need for further discussion. I'll guarantee that it'll be merged as long as it works.

The included project file doesn't work because it just has `antora/antora` for `repo`. Any updates you push should build and deploy right away, I did add the GitLab key to the Netlify site.

mojavelinux commented:
Thanks for the clarifications. That's the direction / commitment I needed. I didn't mean to sound like I know better, just that I need to know what you expect. And you've made that clear. I will proceed as proposed and let you know when it's ready. Cheers!

mojavelinux commented:
> The included project file doesn't work because it just has antora/antora for repo

:man_facepalming:

(I'll make sure the fields are actually right after the next round of updates).

mojavelinux commented:
> I still can't build the site locally without upgrading React. I'm not sure how it's working on Netlify.

I now understand now why this is happening. The yarn.lock and package-lock.json don't agree with each other. I was installing using `npm i`, which gives me react 16.2.0. If I install with `yarn`, I get 16.3.2 (which is what is being used on Netlify). 16.3 works. 16.2 doesn't work.

If I get a chance, I'll address that situation in a separate PR.

mojavelinux commented:
I believe I have implemented the changes as requested.

I also updated the README to including information about the GitLab token. While in there, I mentioned that the GitHub token requires both the repo and gist scopes.

For some reason, my updates are not triggering the deploy preview on Netlify. I'm seeing "Deploy Preview #511: issue-510-gitlab-support@0be6743 skipped" at https://app.netlify.com/sites/staticgen/deploys

mojavelinux commented:
Although Antora doesn't yet make use of it, it's also possible to specify the startertemplaterepo as:

```
startertemplaterepo: antora/demo/docs-site
```

The host is assumed to be the same as repohost unless the value contains a colon (indicating it's some sort of git URL).

mojavelinux commented:
@erquhart I've updated the code. Could you take a look when you get a chance?

erquhart commented:
Sorry for the delay - preview looks good, let's get it in. Thanks so much for your efforts and patience! Also note that we won't see live data from GitLab until the next cache refresh, some time in the next 24 hours.

erquhart commented:
Update: looks like we're passing just the repo (Eg. `"antora/antora"`) throughout the site instead of the url, so the repo links don't work. I rolled the published site back to the previous deploy and locked it there for now. Let me know if you had a plan for finishing that out, I probably should have waited for you to indicate it was merge ready.

mojavelinux commented:
> looks like we're passing just the repo (Eg. "antora/antora") throughout the site instead of the url, so the repo links don't work.

I swear I tested that. But then again, we've changed this so many times, it must have slipped through the cracks. I'll retest.

mojavelinux commented:
> looks like we're passing just the repo (Eg. "antora/antora") throughout the site instead of the url, so the repo links don't work.

Actually, this is incorrect. As I said, the repo in the cache needs to be the URL, which is what fetch-archive now stores. Here's a sample:

```json
{"timestamp":1560500313364,"data":{"antora":[{"timestamp":1560500313364,"followers":540,"stars":153,"forks":60,"issues":114,"repo":"https://gitlab.com/antora/antora"}],"jekyll":[{"timestamp":1560500313364,"followers":6804,"stars":37958,"forks":8278,"issues":140,"repo":"https://github.com/jekyll/jekyll"}]}}
```

Here's how it looks after I run the build: https://gist.github.com/mojavelinux/06cb0318e4fb23d319be277f569f530e

I even put up a sample Netlify site with this code to prove that if you flush the cache, it will work. You can see it in action here: 

* https://staticgen-gitlab-support.netlify.com/antora/
* https://staticgen-gitlab-support.netlify.com/jekyll/

So what doesn't work exactly?

mojavelinux commented:
Also, why do the links from the cards on the home page hard code the domain? That seems unnecessary (and makes testing confusing).

mojavelinux commented:
Here's where the repo from the cache gets assigned to the model:

* https://github.com/netlify/staticgen/blob/master/static.config.js#L70
* https://github.com/netlify/staticgen/blob/master/static.config.js#L89

mojavelinux commented:
It turns out, it might be a completely separate problem. I think the gist download broke, perhaps because something changed in GitHub's API. On a second run, my local archive is always corrupt. As a result, all the links, stars, and followers are missing because there's no data to read.

mojavelinux commented:
I can confirm this is the problem. I'll try to figure out why it's happening, though likely I'll file a new issue.

mojavelinux commented:
See #515

mojavelinux commented:
Btw, thanks for merging!

erquhart commented:
No problem - thanks for following up with the truncation fix!
